### DISCUSSION TITLE
본인이 생각하는 상태 패턴의 장단점에 대해 설명해주세요.

### DISCUSSION THREAD

**Discussion by jimi567** [2024-woowacourse-study/level-interview#55] [OOP]
.

**Comment by kunsanglee** [comment]
상태(State) 패턴은 객체의 상태 변화에 따라 객체의 행동을 유연하게 변경할 수 있도록 하는 디자인 패턴입니다. 

### 장점
- 상태에 따른 행동을 상태 객체 내에서 정의할 수 있으므로, 조건문의 사용이 줄어들고, 각 상태를 별도의 클래스로 관리할 수 있습니다.
- 새로운 상태를 추가해야 할 경우 상태 인터페이스를 구현하는 새로운 클래스를 추가하기만 하면 되므로, 기존 코드를 수정하지 않고도 쉽게 상태를 추가할 수 있습니다.
- 각 상태 클래스 내에서 다음 상태로의 전이를 명확하게 정의할 수 있어, 상태 전이 로직이 한 곳에 집중되고 명확해집니다.
- 상태에 따른 행동을 상태 클래스 내에 캡슐화함으로써, 높은 응집도와 낮은 결합도를 달성할 수 있습니다.

### 단점
- 상태 패턴의 사용은 각 상태를 별도의 클래스로 구현해야 하기 때문에, 상태의 수만큼 클래스가 필요합니다. 이는 프로젝트의 복잡성을 증가시킬 수 있습니다. 
- 상태 객체의 생성과 관리에 따른 추가적인 오버헤드가 발생할 수 있습니다. 특히, 상태 전환의 빈도가 높거나 상태 객체가 많을 경우 이러한 오버헤드가 두드러질 수 있습니다.

상태 패턴은 상태의 수가 많고 복잡한 시스템에서 장점이 잘 발휘되지만, 상태 관리가 비교적 간단한 경우에는 과도한 설계가 될 수 있습니다. 따라서 상태 패턴의 적용 여부를 결정할 때는 시스템의 요구 사항과 복잡성을 면밀히 고려해야 합니다.

**Comment by seunghye218** [comment]
상태 패턴은 객체가 내부 상태에 따라 동작을 변경할 수 있도록 하는 디자인 패턴입니다. 다양한 상태를 별도의 객체로 캡슐화하고 동작을 객체로 표현합니다.

상태 패턴을 적용함으로써 상태의 각 단계에 대해 별도의 객체를 만들 수 있습니다. 각 상태 개체는 해당 단계에 대한 특정 논리와 규칙을 캡슐화하여 코드를 더욱 모듈화하고 유지 관리 가능하며 이해하기 쉽습니다.

객체의 동작이 내부 상태에 따라 달라져야 하는 유사한 사용 사례에 적용하기 좋습니다.

**Comment by GIVEN53** [comment]
조건문 제거
상태 확장 용이
행동 캡슐화

상태마다 클래스 구현
상태 생성, 관리의 오버헤드
